API data-types & specifications for arbitrary metadata signing.

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

ADP-481 > ADP-508

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 209781bd4b5aee3df596ec05ae2b8cd7e955e913
  :round_pushpin: **define ToText/FromText instances for 'DerivationIndex'**
    I had to also move it in '...Primitive.AddressDerivation' because of otherwise circular dependencies (since this module is using the Bounded instances of 'Index', relying on the 'DerivationType' as type parameter. This is mostly extracting the text parsing that was already done in the JSON instance, and moving it as a ToText/FromText instances (and re-using it to build the JSON instances). I've however added a quick sanity text roundtrip property test in the meantime.

- c96593646a433eeba840eea6aab1c85c9753e61c
  :round_pushpin: **rename UTxO*ternal -> Utxo*ternal**
    We use the data-type constructor in order to generate text instances.
  However, our automatic casing function is doing pretty bad with the
  mixed cases and we end up with something like:

    - u_tx_o_external
    - u_tx_o_internal
    - mutable_account

  Instead of fixing the instance itself to work around this, I found it
  easier to just fix the type (which isn't used in many places).

- c4cbafe71b3c9a4271afb8c4417a85f1b8fd8171
  :round_pushpin: **do not output back invalid input when parsing text values.**
    This is a rather bad practice; if a value failed to parse, then we have very little idea about _what_ it is and what's it's shape. It may not be even readable / printable. Instead, the error message should show valid expected options.

- fb5ab0d15d1c37a1b70c2fd4caae7c86f4bc78ce
  :round_pushpin: **extend API and specifications to include the 'signMetadata' operation**

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
define ToText/FromText instances for 'DerivationIndex'

  I had to also move it in '...Primitive.AddressDerivation' because of otherwise circular dependencies (since this module is using the Bounded instances of 'Index', relying on the 'DerivationType' as type parameter. This is mostly extracting the text parsing that was already done in the JSON instance, and moving it as a ToText/FromText instances (and re-using it to build the JSON instances). I've however added a quick sanity text roundtrip property test in the meantime.

KtorZ committed:
rename UTxO*ternal -> Utxo*ternal

  We use the data-type constructor in order to generate text instances.
  However, our automatic casing function is doing pretty bad with the
  mixed cases and we end up with something like:

    - u_tx_o_external
    - u_tx_o_internal
    - mutable_account

  Instead of fixing the instance itself to work around this, I found it
  easier to just fix the type (which isn't used in many places).

KtorZ committed:
extend API and specifications to include the 'signMetadata' operation

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -976,45 +977,10 @@ instance EncodeAddress n => ToJSON (ApiAddress n) where
-                Just s ->
-                    pure s
+    parseJSON = parseJSON
+        >=> fmap ApiT . eitherToParser . first ShowFmt . fromText
```
Parsing delegated to the `FromText` instance and moved near the data-type declaration.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -976,45 +977,10 @@ instance EncodeAddress n => ToJSON (ApiAddress n) where
-      where
-        firstHardened = getIndex @'Hardened minBound
-
+    toJSON = toJSON . toText . getApiT
```
Serialization delegated to the `ToText` instance and moved near the data-type declaration.

KtorZ commented on specifications/api/swagger.yaml:
```diff
@@ -1706,18 +1715,18 @@ x-parametersWalletId: &parametersWalletId
+    The stake the user intends to delegate in Lovelace. Required.
 
-      > ⚠️  On the incentivized testnet, this parameter is not requred, but rather completely ignored.
+    > ⚠️  On the incentivized testnet, this parameter is not requred, but rather completely ignored.
```
This was somewhat indented too far. Sorry for the noise.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -171,8 +169,8 @@ data Depth = RootK | PurposeK | CoinTypeK | AccountK | RoleK | AddressK
 data AccountingStyle
-    = UTxOExternal
-    | UTxOInternal
+    = UtxoExternal
+    | UtxoInternal
```
I think this could be fixed in `fromTextToBoundedEnum SnakeLowerCase`.
It's spelt UTxO everywhere else.
This important word deserves special case treatment, so that we don't need to mangle our type names just to get the preferred serialisation.

KtorZ replied:
I thought of both options and thought that it was cleaner in the end to rename the type here. It's perhaps a little less pervasive as a change to just deal with it in the toText//fromText instances. Any way requires a database fix / migration / work-around :|

rvl commented on specifications/api/swagger.yaml:
```diff
@@ -2293,8 +2331,33 @@ x-tagGroups:
+    - Experimental
 
 paths:
+  /wallets/{walletId}/{role}/{index}/signatures:
```
:+1:

rvl reviewed (approved):
:+1:

KtorZ committed:
do not output back invalid input when parsing text values.

  This is a rather bad practice; if a value failed to parse, then we have very little idea about _what_ it is and what's it's shape. It may not be even readable / printable. Instead, the error message should show valid expected options.

rvl committed:
Fix integration test assertion

rvl committed:
hlint
